puffin/src/scope_details.rs

```diff
@@ -0,0 +1,212 @@
+    pub scope_id: ScopeId,
+    // Custom provided static id that identifiers a scope in a function.
+    pub scope_name: &'static str,
+    /// Scope name, or function name (previously called "id")
+    pub raw_function_name: &'static str,
+    /// Path to the file containing the profiling macro
+    pub file: &'static str,
+    /// The line number containing the profiling
+    pub line_nr: u32,
+}
+
+/// Custom scope details that can be registered by external users.
+/// Instantiate this type once for each custom scope that you record manually.
+/// Custom scopes can be registered via `GlobalProfiler::scope_details().insert_scopes()`.
+// This type provides slightly more convenient api for external users.
+#[derive(Debug, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+pub struct CustomScopeDetails {
+    /// Unique identifier for this scope.
+    pub scope_name: Cow<'static, str>,
+    /// Scope name, or function name (previously called "id")
+    pub function_name: Cow<'static, str>,
+    /// Path to the file containing the profiling macro
+    /// Can be empty if unused.
+    pub file_name: Cow<'static, str>,
+    /// The line number containing the profiling.
+    /// Can be 0 if unused.
+    pub line_nr: u32,
+}
+
+impl CustomScopeDetails {
+    /// Create a new custom scope with a unique name.
+    pub fn new(scope_name: impl Into<Cow<'static, str>>) -> Self {
+        Self {
+            scope_name: scope_name.into(),
+            function_name: Default::default(),
+            file_name: Default::default(),
+            line_nr: Default::default(),
+        }
+    }
+
+    pub fn with_function_name(mut self, name: impl Into<Cow<'static, str>>) -> Self {
+        self.function_name = name.into();
+        self
+    }
+
+    pub fn with_file(mut self, file: Cow<'static, str>) -> Self {
+        self.file_name = file;
+        self
+    }
+
+    pub fn with_line_nr(mut self, line_nr: u32) -> Self {
+        self.line_nr = line_nr;
+        self
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+/// This struct contains scope details and can be read by external libraries.
+pub struct ScopeDetailsOwned {
+    pub scope_name: Cow<'static, str>,
+    /// Shorter variant of the raw function name.
+    /// This is more descriptive and shorter.
+    /// In the case of custom scopes provided scopes this contains the `scope name`.
+    pub dynamic_function_name: Cow<'static, str>,
+    /// Raw function name with the entire type path.
+    /// This is empty for custom scopes.
+    pub raw_function_name: &'static str,
+    /// Shorter variant of the raw file path.
+    /// This is cleaned up and made consistent across platforms.
+    /// In the case of custom scopes provided scopes this contains the `file name`.
+    pub dynamic_file_path: Cow<'static, str>,
+    /// The full raw file path to the file in which the scope was located.
+    /// This is empty for custom scopes.
+    pub raw_file_path: &'static str,
```

Comment:
How can we use `&'static` in a struct `ScopeDetailsOwned`? Aren't the raw file paths logged?